Edit Applications

bengolder opened the pull request:
Closes #1213 
Closes #1214 
Closes #1224
Closes #1248

### What does this do and why?

This adds a way for logged in users (at partner orgs or at CfA) to edit certain data on applications, so that everyone can more easily make corrections to existing applications.

### Testing & Checks

What does this PR include?
- [ ] new dependencies
- [ ] migrations
- [ ] data migrations
- [ ] celery tasks
- [ ] changes to environment variables or local settings
- [ ] configuration changes for 3rd party services (Travis, Heroku, AWS, Github, etc) 
- [ ] visual or style changes

Are there any human checks that should be used to verify these changes, such as user acceptance tests (UAT) or manual deployment checks?

_Deployment Checks_
- [ ] run `manage.py backfill_create_crudevents` on `cmr-stage` to ensure that the command works correctly.
- [ ] run ^ the same command again on prod once we've deployed.

### UAT

#### ORG USER LOGIN EMAIL

Emails are now `cmrtestuser+{org}@gmail.com` for login. For example:
- `cmrtestuser+cfa@gmail.com`
- `cmrtestuser+san_diego_pubdef@gmail.com`

_UAT Checks_
- [x] Edit an application and save the results. Make sure the new data is displayed in the app detail page. Make sure notification emails/text messages are received as org user and as applicant (org user address is now cmrtestuser+{org}@gmail.com)
- [x] Open [an application with existing bad data](https://cmr-develop-pr-1228.herokuapp.com/application/40/). Note that as an org user you cannot submit with errors
- [ ] Read the messages in the comments below, make sure that the writing is satisfactory

Any integration or unit tests?

This adds a new behave integration test for the main editing workflow as well as a set of unit tests for the new `AppEditView`. Unit tests for the `edit_form_service` are coming soon.

### Details

**Editing apps that already have bad data**

Applications that already have bad data are a small percentage of total apps (2%). Most of these have only one invalid field. And an even smaller fraction are likely to be edited by anyone. In the unlikely case that someone edits an application with existing bad data, they will see errors displayed for any invalid fields. They will not be able to submit without correcting the invalid field.

**Notifications to applicants**

- _Skipping contact preference checks._ When we send notifications to applicants regarding an edit to contact information, we do not filter the notifications based on contact preferences.
- _Don't filter for valid contact info._ When we send out applicant edit notifications, we do not check to ensure that the contact info is valid before trying to send a message. This means that in the few cases where an edit was made and some of the involved contact info is invalid, Front will get a message back from the email or phone provider saying that there was a problem delivering the message.

bengolder commented:
**Example message to new phone number**
![img_4060](https://user-images.githubusercontent.com/451510/31640683-9c0c6fda-b294-11e7-8851-58261d308782.PNG)

**Example text to old phone number**
![img_4062](https://user-images.githubusercontent.com/451510/31640716-cc070ede-b294-11e7-9cc3-2ca3ef9637ab.PNG)

**Example with a bunch of changed fields**
![img_4063](https://user-images.githubusercontent.com/451510/31640754-04c41032-b295-11e7-96f3-3c12aade69bd.PNG)

lkogler commented:
Sample emails:
```
Hello,

Your application has been updated by the San Diego County Public Defender.
Here is the information that was changed:
- Phone number

You are receiving this email because Clear My Record has an application on file with this email address.

We value your privacy. If you did not approve this change or would like more information, please email Fresh.Start@sdcounty.ca.gov to contact the San Diego County Public Defender.
```

```
Hello,

Your application has been updated by Code for America.
Here is the information that was changed:
- Address
- Date of birth
- Email
- Phone number

You are receiving this email because Clear My Record has an application on file with this email address.
This is the last email you will get from us at this email address.

We value your privacy. If you did not approve this change or would like more information, please call (415) 301-6005 or email clearmyrecord@codeforamerica.org to contact Code for America.
```

bengolder commented:
## Notifications for org users

**Subject lines include applicant ID**
```
Clear My Record: Updated info for applicant (145)
```

**Example with an address**

```
The information for applicant Jeff Walker has been edited by bgolder+demo+cfa_user@codeforamerica.org from Code for America. You may want to update your organization's records.

------------
Updated Phone number

Before:
(415) 212-4848

After:
(831) 420-7603

------------
Updated Email

Before:
cmrtestuser@gmail.com

After:
cmrtestuser+newemail@gmail.com

------------
Updated Address

Before:
9545 Reyes Mountains Apt. 833
Karenland, AL
92922

After:
9545 Reyes Mountains Apt. 833
Karen, AL
92922

------------
Updated Date of birth

Before:
102810/311/1993

After:
1/1/1993

You can view Jeff Walker's full application here:
https://localhost:8000/application/45/


If you have any questions, please feel free to email us at:
clearmyrecord@codeforamerica.org
```


**Example with a blank data field**

```
The information for applicant Foo Bar has been edited by bgolder+demo+fresno_pubdef_user@codeforamerica.org from Fresno County Public Defender. You may want to update your organization's records.

------------
Updated First name

Before:
Deanna

After:
Foo

------------
Updated Last name

Before:
Fowler

After:
Bar

------------
Updated Phone number

Before:
(Blank)

After:
(415) 301-6005

You can view Foo Bar's full application here:
https://localhost:8000/application/46/


If you have any questions, please feel free to email us at:
clearmyrecord@codeforamerica.org
```

**Example with a sensitive data field (Driver's license)**
```
The information for applicant Foo Bar has been edited by bgolder+demo+fresno_pubdef_user@codeforamerica.org from Fresno County Public Defender. You may want to update your organization's records.

------------
Updated First name

Before:
Andrew

After:
Foo

------------
Updated Last name

Before:
Murphy

After:
Bar

------------
Updated Phone number

Before:
(415) 212-4848

After:
(415) 301-6005

------------
Updated Email

Before:
cmrtestuser@gmail.com

After:
something@example.horse

------------
Updated Driver License/ID



You can view Foo Bar's full application here:
https://localhost:8000/application/50/


If you have any questions, please feel free to email us at:
clearmyrecord@codeforamerica.org
```
